### PR TITLE
[Fix] "Your business location does not match your store location" warning on Payments setting page

### LINKED ISSUE #67461: WooCommerce 11.0.0: False "Your business location does not match your store location" warning on Payments setting page
### Prerequisites

- [x] I have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

After upgrading from WooCommerce 10.7.x to 11.0.0, the Payments settings page incorrectly displays: "Your business location does not match your store location." even though the store location is configured correctly. The issue is reproducible on multiple sites and does not occur in WooCommerce 10.7.x.

### Expected behavior

When the store location is configured correctly (e.g. Canada), the Payments settings page should not display the location mismatch warning.

`window.wcSettings.admin.preloadSettings.general` should contain the General settings, including `woocommerce_default_country`, allowing the business location check to use the configured store location instead of falling back to the default.

### Actual behavior

After upgrading from WooCommerce 10.7.x to 11.0.0, the Payments settings page incorrectly displays the business location mismatch warning, even though the store location is correctly configured.

The issue occurs because `window.wcSettings.admin.preloadSettings.general` is an empty object ({}), causing the following JavaScript to fall back to "US":

```
(
    window.wcSettings?.admin?.preloadSettings?.general
        ?.woocommerce_default_country || 'US'
).split(':')[0]
```

During investigation, `WC_REST_Setting_Options_V2_Controller::get_group_settings( 'general' )` returns a `WP_Error` with the message "Invalid setting group.", which results in the empty preload data.

### Steps to reproduce

1. Install WooCommerce 10.7.x.
2. Configure the store location (e.g. Canada).
3. Upgrade to WooCommerce 11.0.0.
4. Open WooCommerce > Settings > Payments.
5. Click on the yellow warning icon in the "Business Location" dropdown.

Image

### Additional context

The warning originates from JavaScript:

```
(
    window.wcSettings?.admin?.preloadSettings?.general
        ?.woocommerce_default_country || 'US'
).split(':')[0]
```

However, `window.wcSettings.admin.preloadSettings.general` is empty. Therefore the code falls back to "US".

Further investigation shows that

```
$controller = new WC_REST_Setting_Options_V2_Controller();

$result = $controller->get_group_settings( 'general' );
```

returns `WP_Error`: "Invalid setting group."

### WordPress Environment

<details>

<summary>WordPress Environment</summary>

- WordPress: 7.0.2
- WooCommerce: 11.0.0
- PHP: 8.x
- Theme: Storefront
- No custom plugins.
- Store location: Canada.

The issue is reproducible on a clean WooCommerce installation after upgrading from 10.7.x to 11.0.0.

</details>

### Isolating the problem

- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/products/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by kalessil**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes #67461
Related #54638, #65505

Explicitly initialize the admin setting via `admin_init` (restore accidental coupling behavior).

As much as I'm favoring the boyscout rule, this fix must not go that route:
- we have a bad case of accidental coupling
- the fix testability raises another red flag: UTs framework now allows testing a proper split between admin init and Woo core init, forcing us to place both hook initializations in Woo core init

 That's a high-risk profile, and no cleanup should be applied (whatever AIs say) - minimal code changes only.
And yes, all I do here is place a "landmine flag"; I don't have context ("equipment") to safely navigate this minefield.

Now tradeoffs and impact:
- #65505: not affected, setting initialization is low cost compared to the optimization itself
- potential double-loading: low performance impact, low probability ([the scenario](https://github.com/woocommerce/woocommerce/pull/67494#issuecomment-5222325414) was verified agains `all-plugins` none are executed that scenario) 

### How to test the changes in this Pull Request:

Pre-requisite: add the following snippet as a mu-plugin (or to any other place where it'll be executed):

```php
// AI slop, just for testing
add_action( 'admin_bar_menu', function ( $wp_admin_bar ) {
	if ( ! current_user_can( 'manage_woocommerce' ) ) {
		return;
	}

	$address_parts       = array();
	$settings_controller = new WC_REST_Setting_Options_V2_Controller();
	$general             = $settings_controller->get_group_settings( 'general' );

	if ( ! is_wp_error( $general ) ) {
		$fields = array(
			'woocommerce_store_address',
			'woocommerce_store_address_2',
			'woocommerce_store_city',
			'woocommerce_store_postcode',
			'woocommerce_default_country',
		);
		$values = array();
		foreach ( $general as $option ) {
			if ( isset( $option['id'], $option['value'] ) ) {
				$values[ $option['id'] ] = $option['value'];
			}
		}
		foreach ( $fields as $field ) {
			$val = trim( $values[ $field ] ?? '' );
			if ( '' !== $val ) {
				$address_parts[] = $val;
			}
		}
	}

	$location = ! empty( $address_parts ) ? implode( ', ', $address_parts ) : '(unavailable)';

	$wp_admin_bar->add_node( array(
		'id'    => 'wc-settings-monitor',
		'title' => esc_html( $location ),
	) );
}, 999 );
```

- Ensure the store address is configured outside of the US
- For each of the following pages: Dashboard, Orders, Products, Settings, Analytics
- Navigate to the page, refresh it (to ensure SPA navigation is not affecting the testing), and verify the address in the top admin bar.
- Feel free to verify any additional pages
- For all verified pages, the address should be correct.

### CODE REVIEW

**Comment by vladolaru** [comment]
Hey Vladimir — nice catch on the root cause, that `rest_api_init` side-effect dependency is genuinely hard to spot. I reproduced the bug and your fix locally (wp-env, WooCommerce-only, store `CA:ON`), so here's what came out of it.

**Confirmed**, by reverse-applying and re-applying your diff:

| | `preloadSettings.general` | store country in JS | warning |
| --- | --- | --- | --- |
| trunk | `{}` | `US` (fallback) | shown |
| this PR | 20 keys, `CA:ON` | `CA` | gone |

True positives still work — business country forced to `GB` against a `CA` store correctly shows the indicator.

**Why reports have been inconsistent:** any plugin that boots REST during an admin request masks this. With WooPayments active the page looked fine; deactivating it made the bug appear instantly. Probably worth a line in the testing instructions so reviewers don't test on a masked setup.

**One gap in the guard.** It only catches `rest_api_init` firing *after* `admin_init`. If something boots REST earlier (`init`, `plugins_loaded`), `did_action( 'admin_init' )` is still `0`, the guard passes, and `admin_init` then registers everything again. Tested with an mu-plugin doing `rest_get_server()` on `init` priority 1, with your fix applied: **63 `woocommerce_settings_groups` callbacks for 32 unique ids, all 31 groups duplicated.**

The guard is definitely doing real work — removing it entirely gives that same 63/32 split with just WooPayments active. It's the condition that needs adjusting. A simple "already registered?" instance flag with an early return would be order-independent:

```php
if ( $this->wp_admin_settings_registered ) {
	return;
}
$this->wp_admin_settings_registered = true;
```

A few smaller things:

- **Wider than Payments.** `Internal/Admin/Homescreen.php` also reads `preloadSettings['general']` for `woocommerce_store_address` and `woocommerce_default_country` (the latter without a null guard). Might be worth framing the changelog as "restore settings group registration on admin pages".
- **Perf trade-off worth a sentence.** `register_wp_admin_settings()` instantiates every settings page and email object, so `admin_init` runs that on every admin request — including `admin-ajax.php`. Since #65505 was performance work, a reviewer will likely ask. A lazy trigger on first use of `woocommerce_settings_groups` would avoid the unconditional cost, if that's not too invasive for something that needs to ship fast.
- **The `WP_Error` is swallowed.** In `Internal/Admin/Settings.php` and `Loader.php`, `foreach` over the `WP_Error` from `get_group_settings()` silently yields an empty array — no warning, no log. That silence is why this shipped. An `is_wp_error()` check seems worth adding either way.
- **No test asserts settings groups are registered on a plain admin request**, which is exactly why #65505 looked safe.
- Failing check is just `Validate changelog`. And `Bug introduced in PR #65505.` would fit the template's traceability line.

Happy to re-run the A/B against whatever you land on — the repro is set up. Thanks for jumping on this so quickly.

File: plugins/woocommerce/includes/class-woocommerce.php [hunk 1]
```diff
@@ -1536,6 +1542,11 @@ public function get_global( string $global_name ) {
 	 * @return void
 	 */
 	public function register_wp_admin_settings() {
+		// Avoid double-loading in admin caused by rest_preload_api_request calls (e.g. analytics page).
+		if ( doing_action( 'rest_api_init' ) && did_action( 'admin_init' ) ) {
+			return;
```
[the remark this answers is not in the document]
Comment (kalessil) [reply plugins/woocommerce/includes/class-woocommerce.php:1547]: Similar concern to https://github.com/woocommerce/woocommerce/pull/67494#issuecomment-5222325414.
See the PR description for why I'm opposing further hardening here.

[the remark this answers is not in the document]
Comment (kalessil) [reply plugins/woocommerce/includes/class-woocommerce.php:1547]: @coderabbitai: Do you agree with my assessment in the PR description?

**Review by oaratovskyi — APPROVED** [review]
Thanks for turning this around over the weekend, Vladimir — approving for 11.0.1. Tested locally (WC 11.1.0-dev, WP 6.9, PHP 8.1, Storefront, 12 plugins incl. WooPayments, store `CA:ON`) and it fixes the reported bug.

Your test plan passes: Dashboard, Orders, Products, Settings and Analytics all show `150 Elgin Street, Ottawa, K2P 1L4, CA:ON`, 20 `preloadSettings.general` keys, `CA` in JS. True positives still fire — business `US` against a `CA` store shows the indicator.

**Your read on the ordering was right.** On my store `admin_init` fires before `rest_api_init` and the guard does its job: 55 settings-group callbacks, 55 unique, 0 duplicated.

**Two hook orderings still break, and neither breaks on trunk.** Each row ran twice on the same store and probe — once on this branch, once with `class-woocommerce.php` alone reverted to the merge-base — so the only variable is this diff.

| scenario | trunk | this PR |
|---|---|---|
| `rest_do_request()` from `admin_init` pri 5 | 55 groups, `general` OK | **2 groups, `WP_Error: "Invalid setting group."`** |
| `rest_get_server()` on `init` pri 1 | 55 cbs / 55 unique / 0 dup | **108 / 55 / 53 dup** |

The first reproduces #67461's own error: `did_action('admin_init')` flips to `1` the moment `admin_init` *starts*, so a REST dispatch from any callback earlier than priority 10 satisfies both halves of the guard and registration is skipped entirely.

Both triggers were mu-plugins I wrote — I couldn't find a real plugin doing either, and neither could your `all-plugins` check. So I read this as "the guard's correctness depends on an ordering nothing enforces", not "this breaks today". Given the shipped bug is certain and these aren't, ship it.

That said, if there's time before the 11.0.1 cut, I'd rather ship the order-independent version, so nobody has to be right about the ecosystem:

```php
if ( $this->wp_admin_settings_registered ) {
	return;
}
$this->wp_admin_settings_registered = true;
```

It's ~10 lines once phpcs wants a `@var` docblock — bigger than your guard, so I'm arguing correctness, not size.

**Worth a line regardless:** the testing instructions don't say the bug is masked by anything booting REST during an admin request. On my store trunk *also* showed 20 preload keys and no warning, so following the steps as written I'd have concluded there was no bug on trunk either. "Deactivate WooPayments first" would save the next reviewer that.

Follow-ups, all fine after the release: the `is_wp_error()` guard at `Settings.php:170` / `Loader.php:378` (that silence is why this shipped), changelog wording (it's wider than Payments — `Homescreen.php` reads the same preload), and the `Bug introduced in PR #65505.` line. The new test asserts hook wiring rather than registration, so it wouldn't catch a recurrence — worth revisiting with the flag. Note `WP_UnitTestCase_Base` restores `$wp_filter` between tests but won't reset a class property, so two behavioural tests would pass alone and fail as a pair.
